crates/sui-core/src/validator_client_monitor/stats.rs

```diff
@@ -0,0 +1,246 @@
+            validator_stats.update_average_latency(feedback.operation, actual_latency);
+            self.update_global_stats(feedback.operation, actual_latency);
+        }
+    }
+
+    /// Update global maximum latency statistics.
+    ///
+    /// For max latencies, we use a special update strategy:
+    /// - If the new latency is higher than the current max, we immediately update to it
+    /// - Otherwise, we apply decay to gradually lower the max over time
+    ///
+    /// This ensures we always capture peak latencies while still allowing the max to decrease
+    /// when network conditions improve to reduce the impact of outliers.
+    pub fn update_global_stats(&mut self, operation: OperationType, latency: Duration) {
+        let latency_secs = latency.as_secs_f64();
+
+        match self.global_stats.max_latencies.entry(operation) {
+            Entry::Occupied(mut entry) => {
+                let current_max = entry.get().get();
+                if latency_secs > current_max {
+                    // New latency is higher - immediately update to this value
+                    entry.get_mut().override_moving_average(latency_secs);
+                } else {
+                    // New latency is lower - apply decay to gradually reduce max
+                    entry.get_mut().update_moving_average(latency_secs);
+                }
+            }
+            Entry::Vacant(entry) => {
+                entry.insert(DecayMovingAverage::new(
+                    latency_secs,
+                    MAX_LATENCY_DECAY_FACTOR,
+                ));
+            }
+        }
+    }
+
+    /// Calculate client-observed scores for all validators.
+    ///
+    /// Returns a map of validator names to their computed scores based on
+    /// client-side observations. Validators that are currently excluded or
+    /// missing required data will not be included in the results.
+    pub fn calculate_all_client_scores(&self) -> HashMap<AuthorityName, f64> {
+        let mut scores = HashMap::new();
+        for (validator, stats) in self.validator_stats.iter() {
+            if let Some(score) = self.calculate_client_score(stats, &self.global_stats) {
+                scores.insert(*validator, score);
+            }
+        }
+
+        scores
+    }
+
+    /// Calculate client-observed score for a single validator.
+    ///
+    /// The score combines reliability and latency metrics as observed by the client,
+    /// weighted according to configuration. Returns None if:
+    /// - The validator is currently in exclusion cooldown period
+    /// - The validator is missing data for any operation type
+    /// - Global stats are missing for normalization
+    ///
+    /// Score calculation:
+    /// 1. Latency scores are normalized against global maximums
+    /// 2. Each operation type has its own weight
+    /// 3. Final score = (weighted_latency_score * latency_weight) + (reliability * reliability_weight)
+    fn calculate_client_score(
```

Comment:
What do you think about not doing any filtering here? Provide all raw scores in ascending order. Allow the monitor itself to handle the processing of the scores. 

Currently it looks like you do this filtering and then the monitor chooses the top k and then appends unscored validators to that list. Maybe you could simplify this flow a bit.

Reply:
Which filtering are you referring to?

Reply:
You are filtering the excluded validators or missing data validators. Just return them with a score of zero and then when you sort you can still select the top k and have the remaining list be used for the fallback